### ISSUE TITLE
Unexpected and Random failure on SSL_Read calls

### ISSUE THREAD

**Issue by tunafish24** [openssl/openssl#10737]
openssl 1.1.1d
TLS 1.3 PSK

I'm working on a server & client application, both use shared ssl codebase to send/recv data over 1 Gbps local ethernet connection. Currently, I'm stress testing the application by sending unlimited amount of data from server and receiving it on the client. The test is being performed on the local network using two computers.

The problem I'm running into is the client application randomly fails on SSL_READ(), it returns -1 and then SSL_get_error() returns 1 (SSL_ERROR_SSL). I printed openssl stack trace and saw following messages on two different test runs.
```
Run 1
3576:error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:ssl\record\ssl3_record.c:677:\n

Run 2
11700:error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:ssl\record\ssl3_record.c:677:\n
```
Also ssl->statem->state = MSG_FLOW_ERROR

These failures occur randomly and after 100s of MBs or 10s of GBs of data has been sent/recv successfully. I'm using Memory BIOs and the logic is relatively straightforward on client/read side.

LOOP 
1. Check if there is pending data using SSL_Read 
2. If SSL_Read failed (SSL_ERROR_WANT_READ) then Read more data from Socket 
3. If more data was read from Socket, then call BIO_Write on the ReadBIO

Im using 16KB buffer for Reading and this works fine for 100s of Megabytes of data = hundred thousands of successful calls. So why does it fail randomly?

On Each side, a single thread handles both SSL_Read & SSL_Write. The Sockets are in non-blocking mode.

Note: Buffer len is always 16KB
```
int Tls::Read(char *buff, int len)
{
    int ret;

    while (true)
    {       

        //Check if SSL_Read works
        ret = SSL_read(ssl, buff, len);

        if (ret > 0)
            return ret; //TLS Record successfully read

        ret = SSL_get_error(ssl, ret);

        if (ret != SSL_ERROR_WANT_READ)         
            return -1;  //TLS Error         


        //Try Reading more from Socket
        bool bytes = pSocket->Read(buff, len);

        if (bytes == 0)
        {
            return 0;   //WouldBlock - Retry
        }

        //Copy Data to Read Bio
        ret = BIO_write(bioRead, buff, bytes);

        if (ret <= 0)
            return -1;
    }
}
```

**Comment by mattcaswell** [comment]
Is this specific to a particular ciphersuite?

**Comment by tunafish24** [comment]
Tried AES256_GCM_SHA384 and CHACHA20_POLY1305_SHA256, both ciphersuites exhibiting same failure with identical error codes & stack traces.

**Comment by mattcaswell** [comment]
Try applying this patch. It won't fix your problem but might (possibly) give us more diagnostics. This patch seems to have an unexpected interaction with early data and is causing one of the early data tests to fail - but assuming you're not using early data that shouldn't be a problem.

If you could run your test again with this patch installed and report back. Make sure your application prints the *whole* error stack if it encounters the problem (i.e. don't just print the top error in the stack). I'm hoping to see an additional error which might give us a clue as to where to look.

```` diff
diff --git a/ssl/record/ssl3_record_tls13.c b/ssl/record/ssl3_record_tls13.c
index ab50e37624..6b58a49a3d 100644
--- a/ssl/record/ssl3_record_tls13.c
+++ b/ssl/record/ssl3_record_tls13.c
@@ -143,6 +143,8 @@ int tls13_enc(SSL *s, SSL3_RECORD *recs, size_t n_recs, int sending)
     }
     if (loop == 0) {
         /* Sequence has wrapped */
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS13_ENC,
+                 ERR_R_INTERNAL_ERROR);
         return -1;
     }
 
@@ -151,6 +153,8 @@ int tls13_enc(SSL *s, SSL3_RECORD *recs, size_t n_recs, int sending)
             || (!sending && EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG,
                                              taglen,
                                              rec->data + rec->length) <= 0)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS13_ENC,
+                 ERR_R_INTERNAL_ERROR);
         return -1;
     }
 
@@ -162,6 +166,8 @@ int tls13_enc(SSL *s, SSL3_RECORD *recs, size_t n_recs, int sending)
             || !WPACKET_get_total_written(&wpkt, &hdrlen)
             || hdrlen != SSL3_RT_HEADER_LENGTH
             || !WPACKET_finish(&wpkt)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS13_ENC,
+                 ERR_R_INTERNAL_ERROR);
         WPACKET_cleanup(&wpkt);
         return -1;
     }
@@ -179,6 +185,8 @@ int tls13_enc(SSL *s, SSL3_RECORD *recs, size_t n_recs, int sending)
                                 (unsigned int)rec->length) <= 0
             || EVP_CipherFinal_ex(ctx, rec->data + lenu, &lenf) <= 0
             || (size_t)(lenu + lenf) != rec->length) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS13_ENC,
+                 ERR_R_INTERNAL_ERROR);
         return -1;
     }
     if (sending) {

````

**Comment by tunafish24** [comment]
So after spending another day trying to figure out the issue, I found the culprit. Turns out there is an extremely rare edge case where 1 byte of outgoing encrypted data would get overwritten by new encrypted data from the server side. That was causing SSL_Read on client to fail. 
Sorry for raising a false alarm. On the plus side, I stress tested the living daylights out of openssl tls 1.3 psk implementation and it survived.
